ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
+				break;
+			case 'domain':
+				// A more restrictive policy.
+				if ($userName !== preg_replace('`[^a-zA-Z0-9 .-]+`i', '', $userName))
+					return null;
+				break;
+			default:
+				// Just kick them if they have an unusable username.
+				if (preg_match('`[#<>[\]|{}@:]+`', $userName))
+					return null;
+		}
+
+		return $userName;
+	}
+
+	private function updateWikiUser()
+	{
+		$this->userWiki->setEmail($this->FSuser['email_address']);
+		$this->userWiki->mEmailAuthenticated = wfTimestampNow(); 
+		$this->userWiki->setRealName($this->FSuser['real_name']);
+
+		// Check some stuff.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+		{
+			// Convert the smf_member_id over?
+			if ($this->userWiki->getOption('forum_member_id', 0) == 0 && $this->userWiki->getOption('smf_member_id', 0) != 0)
+				$user->setOption('forum_member_id', $this->FSuser['id_member']);
+		}
+
+		$this->userWiki->setOption('forum_last_update_user', time());
+		$this->userWiki->saveSettings();
+	}
+
+	/*
+	 * Set all of our wiki groups
+	*/
+	private function setWikiGroups()
+	{
+		// Wiki Group Name => SMF Group IDS
+		$groupActions = array(
+			'sysop' => $this->FSSettings['AdminGroups'],
+//			'interface-admin' => $this->FSSettings['SuperAdminGroups'],
+			'bureaucrat' => $this->FSSettings['SuperAdminGroups'],
+		);
+
+		// Add in our special groups.
+		foreach ($this->FSSettings['SpecialGroups'] as $smf_group_id => $wiki_group_name)
+		{
+			// Group didn't exist?
+			if (!isset($groupActions[$wiki_group_name]))
+				$groupActions[$wiki_group_name] = array();
+
+			// Add the SMF group into the wiki group.
+			$groupActions[$wiki_group_name][] = $smf_group_id;
+		}
+
+		// Now we are going to check all the groups.
+		foreach ($groupActions as $wiki_group_name => $smf_group_ids)
+		{
+			// They are in the SMF group but not the wiki group?
+			if (
+				array_intersect($smf_group_ids, $this->FSgroups) != array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) == array()
+			)
+				$this->userWiki->addGroup($wiki_group_name);
+			// They are not in the SMF group, but in the wiki group
+			elseif (
+				array_intersect($smf_group_ids, $this->FSgroups) == array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) != array()
+			)
+				$this->userWiki->removeGroup($wiki_group_name);
+		}
+
+		// Did we make any changes?
+		$this->userWiki->setOption('forum_last_update_groups', time());
+		$this->userWiki->saveSettings();
+	}
+
+	/*
+	 * Create a Wiki user
+	*/
+	private function createWikiUser()
+	{
+		$this->userWiki->setName($this->FSuserName);
+		$this->userWiki->setEmail($this->FSuser['email_address']);
+		$this->userWiki->setRealName($this->FSuser['real_name']);
+		$this->userWiki->mEmailAuthenticated = wfTimestampNow();
+
+		$this->userWiki->addToDatabase();
+
+		// Only SMF does this.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+			$user->setOption('forum_member_id', $this->FSuser['id_member']);

```

Comment:
The purpose is to attempt to prevent account takeovers.  Since its possible in SMF to change the account name, this would essentially open up the account on the wiki side to be taken over by another member.  This logic is to stop that.  Although there is no clear way of fixing this from the wiki side other than deleting the account.

At one point I did intend to have logic if it couldn't find a member by the name, to instead go search to find the forum member id, then force a name change.

Reply:
Elkarte doesn't work any differently, I just don't allow account name changes willy nilly on my forums.

There's a lot to unpack with this and this isn't a properly complete solution if you want to address this. The situation depends on software (currently the same for everyone) and to prevent such takeovers it needs to be checked in the validation stage like everything else. 

Should remove this for now, and make a separate PR with a more complete solution.

Reply:
Err, sorry if it wasn't clear. There is no point in gating this behind the name style.

Reply:
If I remove the gating for now, I think we can come back around to improving this.  The idea is sound, just yes I haven't gotten around to fully implanting it.

With the ID as well we can reverse link their wiki profile to their forum profile.  We could also maybe allow a wiki admin to unlink the profile as well, allowing it to be linked by another.

The real problem is how both handle allow changing usernames.  It becomes a problem on both sides.  This was my first step in that direction to allow us to ensure accounts can be changed and relinked correctly.